nomad/jobs_endpoint.go

```diff
@@ -0,0 +1,297 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package nomad
+
+import (
+	"errors"
+	"net/http"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-memdb"
+	"github.com/hashicorp/go-set/v2"
+	"github.com/hashicorp/nomad/acl"
+	"github.com/hashicorp/nomad/nomad/state"
+	"github.com/hashicorp/nomad/nomad/state/paginator"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+func NewJobsEndpoint(s *Server, ctx *RPCContext) *Jobs {
+	return &Jobs{
+		srv:    s,
+		ctx:    ctx,
+		logger: s.logger.Named("jobs"),
+	}
+}
+
+type Jobs struct {
+	srv    *Server
+	ctx    *RPCContext
+	logger hclog.Logger
+}
+
+func (j *Jobs) Statuses(
+	args *structs.JobsStatusesRequest,
+	reply *structs.JobsStatusesResponse) error {
+
+	authErr := j.srv.Authenticate(j.ctx, args)
+	if done, err := j.srv.forward("Jobs.Statuses", args, args, reply); done {
+		return err
+	}
+	j.srv.MeasureRPCRate("jobs", structs.RateMetricList, args)
+	if authErr != nil {
+		return structs.ErrPermissionDenied
+	}
+	defer metrics.MeasureSince([]string{"nomad", "jobs", "statuses"}, time.Now())
+
+	namespace := args.RequestNamespace()
+	// The ns from the UI by default is "*" which scans the whole "jobs" table.
+	// If specific jobs are requested, all with the same namespace,
+	// we may get some extra efficiency, especially for non-contiguous job IDs.
+	if len(args.Jobs) > 0 {
+		nses := set.New[string](0)
+		for _, j := range args.Jobs {
+			nses.Insert(j.Namespace)
+		}
+		if nses.Size() == 1 {
+			namespace = nses.Slice()[0]
+		}
+	}
```

Comment:
This is a nice little optimization, especially when combined with the early return we'd get because of the `AllocNsOp` check on line 69. :+1: